### PR TITLE
[CDN] Add cli support for AFD SKU

### PR DESCRIPTION
**Pull request by t-bzhan**
**Description**

Add cli support for AFD SKU

**Testing Guide**

az afd endpoint
az afd custom-domain
az afd origin-group
az afd origin
az afd route
az afd rule-set
az afd rule
az afd security-policy
az afd secret

**History Notes**

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [x] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [x] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [x] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
CDN

**Comment by t-bzhan** [comment]
@yonzhan , is it possible for this PR to be included in S183 milestone?

File: src/azure-cli/azure/cli/command_modules/cdn/_help.py [hunk 1]
```diff
@@ -987,3 +987,658 @@
...
+"""
+
+helps['afd custom-domain'] = """
+type: group
+short-summary: Manage custom domains within the specified profile.
+"""
+
+helps['afd custom-domain create'] = """
+type: command
+short-summary: Create a custom domain within the specified profile. Your need to create a DNS TXT record to prove the custom domain's ownership to make the operation successful.
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/cdn/_help.py:1249]: This summary is too long. You should move some part of summary into `long-summary`

File: src/azure-cli/azure/cli/command_modules/cdn/_help.py [hunk 2]
```diff
@@ -987,3 +987,658 @@
...
+    short-summary: >
+        The number of samples to consider for load balancing decisions.
+  - name: --load-balancing-successful-samples-required
+    type: int
+    short-summary: >
+        The number of samples within the sample period that must succeed for the orgin to be deemed as healthy.
+  - name: --load-balancing-additional-latency-in-milliseconds
+    type: int
+    short-summary: >
+        The additional latency in milliseconds for probes to fall into the lowest latency bucket.
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/cdn/_help.py:1030]: The name of parameter is too long. Can you add a short name as alias?

File: src/azure-cli/azure/cli/command_modules/cdn/_help.py [hunk 3]
```diff
@@ -987,3 +987,658 @@
...
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/cdn/_help.py:1075]: The help message of parameter can be defined in _params.py.

Comment (kairu-ms) [reply src/azure-cli/azure/cli/command_modules/cdn/_help.py:1075]: I found you already setting the help property of parameter in _params.py. Then you don't need to write them again here

File: src/azure-cli/azure/cli/command_modules/cdn/_params.py [hunk 4]
```diff
@@ -262,3 +221,305 @@ def load_arguments(self, _):
...
+                   help='The number of seconds between health probes.Default is 240 seconds.')
+        c.argument('probe_path',
+                   help="The path relative to the origin that is used to determine the health of the origin.")
+
+        # load balancing related parameters
+        c.argument('load_balancing_sample_size', type=int,
+                   help="The number of samples to consider for load balancing decisions.")
+        c.argument('load_balancing_successful_samples_required', type=int,
+                   help="The number of samples within the sample period that must succeed.")
+        c.argument('load_balancing_additional_latency_in_milliseconds', type=int,
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/cdn/_params.py:270]: please add a short name as the alias in options_list property

File: src/azure-cli/azure/cli/command_modules/cdn/_params.py [hunk 5]
```diff
@@ -262,3 +221,305 @@ def load_arguments(self, _):
...
+    # AFD Origin Group #
+    with self.argument_context('afd origin-group') as c:
+        c.argument('origin_group_name', origin_group_name_type, id_part='child_name_1',
+                   help='Name of the origin group.')
+        c.argument('profile_name', help=profile_name_help, id_part='name')
+
+        # health probe related paramters
+        c.argument('probe_request_type', arg_type=get_enum_type([item.value for item in list(HealthProbeRequestType)]),
+                   help='The type of health probe request that is made.')
+        c.argument('probe_protocol', arg_type=get_enum_type([item.value for item in list(ProbeProtocol)]),
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/cdn/_params.py:258]:
```suggestion
        c.argument('probe_protocol', arg_type=get_enum_type(ProbeProtocol),
```

File: src/azure-cli/azure/cli/command_modules/cdn/_params.py [hunk 6]
```diff
@@ -262,3 +221,305 @@ def load_arguments(self, _):
...
+        c.argument('profile_name', id_part=None)
+
+    # AFD Origin Group #
+    with self.argument_context('afd origin-group') as c:
+        c.argument('origin_group_name', origin_group_name_type, id_part='child_name_1',
+                   help='Name of the origin group.')
+        c.argument('profile_name', help=profile_name_help, id_part='name')
+
+        # health probe related paramters
+        c.argument('probe_request_type', arg_type=get_enum_type([item.value for item in list(HealthProbeRequestType)]),
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/cdn/_params.py:256]:
```suggestion
        c.argument('probe_request_type', arg_type=get_enum_type(HealthProbeRequestType),
```

File: src/azure-cli/azure/cli/command_modules/cdn/_params.py [hunk 7]
```diff
@@ -262,3 +221,305 @@ def load_arguments(self, _):
...
+
+    # AFD Origin #
+    with self.argument_context('afd origin') as c:
+        c.argument('origin_name', origin_name_type, id_part='child_name_2', help='Name of the origin.')
+        c.argument('profile_name', help=profile_name_help, id_part='name')
+        c.argument('origin_group_name', origin_group_name_type, id_part='child_name_1',
+                   help='Name of the origin group which is unique within the endpoint.')
+        c.argument('http_port', type=int, help="The value of the HTTP port. Must be between 1 and 65535.")
+        c.argument('https_port', type=int, help="The value of the HTTPS port. Must be between 1 and 65535.")
+        c.argument('enabled_state', arg_type=get_enum_type([item.value for item in list(EnabledState)]),
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/cdn/_params.py:285]:
```suggestion
        c.argument('enabled_state', arg_type=get_enum_type(EnabledState),
```

File: src/azure-cli/azure/cli/command_modules/cdn/commands.py [hunk 8]
```diff
@@ -207,3 +275,100 @@ def load_command_table(self, _):
...
         g.custom_show_command('show', 'show_waf_rate_limit_rule', client_factory=cf_waf_policy)
+
+    with self.command_group('afd endpoint', cdn_afd_endpoints_sdk, is_preview=True) as g:
+        g.show_command('show', 'get')
+        g.command('list', 'list_by_profile')
+        g.command('purge', 'purge_content', supports_no_wait=True)
+        g.command('delete', 'delete', confirmation=True)
+
+        g.custom_command('update', 'update_afd_endpoint', client_factory=cf_afd_endpoints)
+        g.custom_command('create', 'create_afd_endpoint', client_factory=cf_cdn,
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/cdn/commands.py:286]: Why the client_factory is cf_cdn?

File: src/azure-cli/azure/cli/command_modules/cdn/commands.py [hunk 9]
```diff
@@ -207,3 +275,100 @@ def load_command_table(self, _):
...
+        g.command('list', 'list_by_rule_set')
+        g.custom_command('create', 'create_afd_rule', client_factory=cf_afd_rules)
+        g.command('delete', 'delete', confirmation=True)
+
+    with self.command_group('afd rule condition', cdn_afd_rule_sdk, is_preview=True) as g:
+        g.custom_command('add', 'add_afd_rule_condition', client_factory=cf_afd_rules,
+                         doc_string_source='azure.mgmt.cdn.models#Rule')
+        g.custom_command('remove', 'remove_afd_rule_condition', client_factory=cf_afd_rules,
+                         doc_string_source='azure.mgmt.cdn.models#Rule')
+        g.custom_command('list', 'list_afd_rule_condition', client_factory=cf_afd_rules)
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/cdn/commands.py:327]: Why not use custom_command_type=XXX property in command_group? If you define a custom command type with client_factory = cf_afd_rules, you don't need to write it again in g.custom_command function.

File: src/azure-cli/azure/cli/command_modules/cdn/_params.py [hunk 10]
```diff
@@ -262,3 +221,305 @@ def load_arguments(self, _):
...
+        c.argument('secret_version', help='Version of the certificate to be used.')
+
+    with self.argument_context('afd secret list') as c:
+        c.argument('profile_name', id_part=None)
+        c.argument('secret_name', id_part=None)
+
+    # AFD Log Analytic #
+    with self.argument_context('afd log-analytic') as c:
+        c.argument('group_by', nargs='+',
+                   arg_type=get_enum_type(["httpStatusCode", "protocol", "cacheStatus", "country", "customDomain"]),
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/cdn/_params.py:426]: Is the enum value defined in SDK? If so you can import the enum class here.

File: src/azure-cli/azure/cli/command_modules/cdn/_params.py [hunk 11]
```diff
@@ -262,3 +221,305 @@ def load_arguments(self, _):
...
+    with self.argument_context('afd security-policy list') as c:
+        c.argument('profile_name', id_part=None)
+        c.argument('security_policy_name', id_part=None)
+
+    # AFD Custom Domain #
+    with self.argument_context('afd custom-domain') as c:
+        c.argument('profile_name', help=profile_name_help, id_part='name')
+        c.argument('custom_domain_name', custom_name_arg_type, id_part="child_name_1",
+                   help='Name of the custom domain.')
+        c.argument('certificate_type', arg_type=get_enum_type([item.value for item in list(AfdCertificateType)]),
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/cdn/_params.py:398]:
```suggestion
        c.argument('certificate_type', arg_type=get_enum_type(AfdCertificateType),
```

File: src/azure-cli/azure/cli/command_modules/cdn/_params.py [hunk 12]
```diff
@@ -262,3 +221,305 @@ def load_arguments(self, _):
...
+                   help='Defines the source of the SSL certificate.')
+        c.argument('minimum_tls_version', arg_type=get_enum_type([item.value for item in list(AfdMinimumTlsVersion)]),
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/cdn/_params.py:400]:
```suggestion
        c.argument('minimum_tls_version', arg_type=get_enum_type(AfdMinimumTlsVersion),
```

File: src/azure-cli/azure/cli/command_modules/cdn/linter_exclusions.yml [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+            rule_exclusions:
+            - option_length_too_long
+afd rule action add:
+    parameters:
+        preserve_unmatched_path:
+            rule_exclusions:
+            - option_length_too_long
+        query_string_behavior:
+            rule_exclusions:
+            - option_length_too_long
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/cdn/linter_exclusions.yml:108]: Can you add short name for them?

File: src/azure-cli/azure/cli/command_modules/cdn/_params.py (hunk 5 above)
Comment (t-bzhan) [reply src/azure-cli/azure/cli/command_modules/cdn/_params.py]: Thanks a lot, fixed.

File: src/azure-cli/azure/cli/command_modules/cdn/commands.py (hunk 8 above)
Comment (t-bzhan) [reply src/azure-cli/azure/cli/command_modules/cdn/commands.py]: Fixed

File: src/azure-cli/azure/cli/command_modules/cdn/_params.py (hunk 4 above)
Comment (t-bzhan) [reply src/azure-cli/azure/cli/command_modules/cdn/_params.py]: I've trip the leading "load_balancing" in the options_list, but it is still a little longer than the limit.

I could not find a suitable short name for those parameters, do you have any suggestion for that? :)

File: src/azure-cli/azure/cli/command_modules/cdn/_params.py (hunk 10 above)
Comment (t-bzhan) [reply src/azure-cli/azure/cli/command_modules/cdn/_params.py:448]: Not yet. This version of swagger definition the SDK is based has some mistakes which omit these enum definition. I will fix it once we have the new SDK generated.

File: src/azure-cli/azure/cli/command_modules/cdn/linter_exclusions.yml (hunk 13 above)
Comment (t-bzhan) [reply src/azure-cli/azure/cli/command_modules/cdn/linter_exclusions.yml:108]: Trip the leading "load_balancing" for those load balancing related ones.

I could not find a suitable short name for other parameters, let me know if you have any ideas? :)

File: src/azure-cli/azure/cli/command_modules/cdn/_help.py [hunk 14]
```diff
@@ -987,3 +996,540 @@
...
+short-summary: List all the endpoints within the specified profile.
+examples:
+  - name: List all the endpoints within the specified profile.
+    text: >
+        az afd endpoint list -g group --profile-name profile
+"""
+
+helps['afd route'] = """
+type: group
+short-summary: Manage routes under an AFD endpoint. Routes are used to associate doamins and origin groups.
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/cdn/_help.py:1245]: I think it's better to use long-summary here?

File: src/azure-cli/azure/cli/command_modules/cdn/_help.py [hunk 15]
```diff
@@ -987,3 +996,540 @@
...
+        az afd route update -g group --endpoint-name endpoint1 --profile-name profile --route-name route1 --rule-sets ruleset1 rulseset2
+  - name: Update a route's compression settings to enable compression for the specified content types.
+    text: >
+        az afd route update -g group --endpoint-name endpoint1 --profile-name profile --route-name route1 --query-string-caching-behavior IgnoreQueryString
+        --enable-compression true --content-types-to-compress text/javascript text/plain
+"""
+
+helps['afd security-policy'] = """
+type: group
+short-summary: Manage security policies within the specified profile. Security policies could be used to apply a web application firewall policy to
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/cdn/_help.py:1286]: I think it's better to use long-summary here?

File: src/azure-cli/azure/cli/command_modules/cdn/_help.py [hunk 16]
```diff
@@ -987,3 +996,540 @@
...
+short-summary: Delete an existing route within the specified endpoint.
+examples:
+  - name: Delete an route named route1.
+    text: >
+        az afd route delete -g group --profile-name profile --endpoint-name endpoint1 --route-name route1
+"""
+
+helps['afd secret'] = """
+type: group
+short-summary: Manage secrets within the specified profile. Secrets are used to reference your own certificate stored in Azure Key Vault.
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/cdn/_help.py:1324]: I think it's better to use long-summary here.

File: src/azure-cli/azure/cli/command_modules/cdn/custom/custom_afdx.py [hunk 17]
```diff
@@ -0,0 +1,829 @@
...
+                               security_policy_name,
+                               domains: List[str],
+                               waf_policy: str):
+
+    if any([("/afdendpoints/" not in domain.lower() and
+             "/customdomains/" not in domain.lower()) for domain in domains]):
+        raise CLIError('Domain should either be endpoint ID or custom domain ID.')
+
+    if "/frontdoorwebapplicationfirewallpolicies/" not in waf_policy.lower():
+        raise CLIError('waf_policy should be valid Front Door WAF policy ID.')
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/cdn/custom/custom_afdx.py:656]: CLIError will be deprecated. Please use error types defined in this doc:
https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md#available-error-types

File: src/azure-cli/azure/cli/command_modules/cdn/custom/custom_afdx.py [hunk 18]
```diff
@@ -0,0 +1,829 @@
...
+                               security_policy_name,
+                               domains: List[str] = None,
+                               waf_policy: str = None):
+
+    if domains is not None and any([("/afdendpoints/" not in domain.lower() and
+                                     "/customdomains/" not in domain.lower()) for domain in domains]):
+        raise CLIError('Domain should either be endpoint ID or custom domain ID.')
+
+    if waf_policy is not None and "/frontdoorwebapplicationfirewallpolicies/" not in waf_policy:
+        raise CLIError('waf_policy should be Front Door WAF policy ID.')
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/cdn/custom/custom_afdx.py:683]: CLIError will be deprecated.

File: src/azure-cli/azure/cli/command_modules/cdn/custom/custom_afdx.py [hunk 19]
```diff
@@ -0,0 +1,829 @@
...
+                      secret_name,
+                      secret_source,
+                      secret_version: str = None,
+                      use_latest_version: bool = None):
+
+    if "/certificates/" not in secret_source.lower():
+        raise CLIError('secret_source should be valid Azure key vault certificate ID.')
+
+    if secret_version is None and not use_latest_version:
+        raise CLIError('Either specify secret_version or enable use_latest_version.')
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/cdn/custom/custom_afdx.py:712]: CLIError will be deprecated.

File: src/azure-cli/azure/cli/command_modules/cdn/custom/custom_afdx.py [hunk 20]
```diff
@@ -0,0 +1,829 @@
...
+                             minimum_tls_version: AfdMinimumTlsVersion,
+                             azure_dns_zone: str = None,
+                             secret: str = None,
+                             no_wait: bool = None):
+
+    if azure_dns_zone is not None and "/dnszones/" not in azure_dns_zone.lower():
+        raise CLIError('azure_dns_zone should be valid Azure dns zone ID.')
+
+    if secret is not None and "/secrets/" not in secret.lower():
+        raise CLIError('secret should be valid Azure Front Door secret ID.')
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/cdn/custom/custom_afdx.py:779]: CLIError will be deprecated.

File: src/azure-cli/azure/cli/command_modules/cdn/custom/custom_afdx.py [hunk 21]
```diff
@@ -0,0 +1,829 @@
...
+                             resource_group_name: str,
+                             profile_name: str,
+                             custom_domain_name: str,
+                             certificate_type: AfdCertificateType = None,
+                             minimum_tls_version: AfdMinimumTlsVersion = None,
+                             azure_dns_zone: str = None,
+                             secret: str = None):
+
+    if azure_dns_zone is not None and "/dnszones/" not in azure_dns_zone.lower():
+        raise CLIError('azure_dns_zone should be valid Azure dns zone ID.')
```
Comment (kairu-ms) [inline src/azure-cli/azure/cli/command_modules/cdn/custom/custom_afdx.py:802]: CLIError will be deprecated.

File: src/azure-cli/azure/cli/command_modules/cdn/custom/custom_afdx.py (hunk 17 above)
Comment (t-bzhan) [reply src/azure-cli/azure/cli/command_modules/cdn/custom/custom_afdx.py:656]: Thanks a lot, will use InvalidArgumentValueError in next iteration.

File: src/azure-cli/azure/cli/command_modules/cdn/_help.py (hunk 15 above)
Comment (t-bzhan) [reply src/azure-cli/azure/cli/command_modules/cdn/_help.py]: Agree, thanks a lot.

**Review by kairu-ms — APPROVED** [review]

File: src/azure-cli/azure/cli/command_modules/cdn/commands.py [hunk 22]
```diff
@@ -207,3 +285,117 @@ def load_command_table(self, _):
...
+        g.custom_command('create', 'create_afd_profile')
+        g.custom_command('usage', 'list_resource_usage', client_factory=cf_afd_profiles)
+
+    with self.command_group('afd endpoint', cdn_afd_endpoints_sdk, is_preview=True) as g:
+        g.show_command('show', 'get')
+        g.command('list', 'list_by_profile')
+        g.command('purge', 'purge_content', supports_no_wait=True)
+        g.command('delete', 'delete', confirmation=True)
+
+        g.custom_command('update', 'update_afd_endpoint', client_factory=cf_afd_endpoints)
```
Comment (jsntcy) [inline src/azure-cli/azure/cli/command_modules/cdn/commands.py:305]: Unnecessary to add `client_factory=cf_afd_endpoints` for g.custom_command as `cdn_afd_endpoints_sdk` in self.command_group already defines client_factory as `cf_afd_endpoints`; other commands as well.

File: src/azure-cli/azure/cli/command_modules/cdn/commands.py [hunk 23]
```diff
@@ -207,3 +285,117 @@ def load_command_table(self, _):
         g.custom_command('delete', 'delete_waf_rate_limit_rule', client_factory=cf_waf_policy, confirmation=True)
         g.custom_command('list', 'list_waf_rate_limit_rules', client_factory=cf_waf_policy)
         g.custom_show_command('show', 'show_waf_rate_limit_rule', client_factory=cf_waf_policy)
+
+    with self.command_group('afd profile', cdn_profiles_sdk,
+                            is_preview=True,
```
Comment (jsntcy) [inline src/azure-cli/azure/cli/command_modules/cdn/commands.py:290]: >is_preview=True [](start = 28, length = 15)

We can define `is_preview=True` for command group `afd`, then all sub command groups will show as preview.

Comment (t-bzhan) [reply src/azure-cli/azure/cli/command_modules/cdn/commands.py]: Thanks, fiexed.

File: src/azure-cli/azure/cli/command_modules/cdn/commands.py (hunk 22 above)
Comment (t-bzhan) [reply src/azure-cli/azure/cli/command_modules/cdn/commands.py]: Thanks, I will move the client_factory parameter to command group level to remove the duplication according to :
image

as my custom command use the client argument.

Comment (jsntcy) [reply src/azure-cli/azure/cli/command_modules/cdn/commands.py]: Unnecessary to add `client_factory=cf_afd_endpoints` again cdn_afd_endpoints_sdk already defines client_factory `cf_afd_endpoints`

---
In reply to: [581681124](https://github.com/Azure/azure-cli/pull/16951#discussion_r581681124) [](ancestors = 581681124)

**Review by jsntcy — APPROVED** [review]
:shipit:

**Pull request merged**